File: ReversedLinkedList.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this : None


// Your code here along with comments explaining your approach

class ReversedLinkedList {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        if(curr==null)
        return curr;
        ListNode fast = curr.next;
        ListNode prev=null;
        while(fast!=null){
            curr.next = prev;
            prev=curr;
            curr=fast;
            fast=fast.next;
        }
        curr.next = prev;
        return curr;
    }
}

File: ReversedLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Time Complexity => O(N)
# Space Complexity => O(1)
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if(head == None):
            return None
        temp = head
        before = None
        while(temp is not None):
            after = temp.next
            temp.next = before
            before = temp
            temp = after
        head = before
        return before
        